core/cache/resident.go

```diff
@@ -0,0 +1,168 @@
+)
+
+// The cached controller includes a "residentManager", which supplies new
+// cache "resident" instances, monitors their life cycles and is the source
+// of unique identifiers for residents and resources.
+// All cached entities aside from the parent controller should include
+// the "resident" type as a base.
+// Each resident monitors resources that it creates and is responsible for
+// cleanup up when it is to be evicted from the cache.
+// This design meets resource management requirements multi-directionally:
+//   1. Resources (such as watchers) that are stopped or otherwise destroyed
+//      by their upstream owners need to be deregistered from their residents.
+//   2. Residents removed from a model in the normal course of events need to
+//      release resources that they created and deregister from the controller.
+//   3. If the multi-watcher supplying deltas to the cache is restarted,
+//      The controller itself must mark and sweep, evicting stale residents and
+//      cleaning up their resources.
+// Where possible the manager and residents eschew responsibility for Goroutine
+// safety. The types into which they are embedded should handle this.
+
+// counter supplies monotonically increasing unique identifiers.
+type counter uint64
+
+func (c *counter) next() uint64 {
+	return atomic.AddUint64((*uint64)(c), 1)
+}
+
+func (c *counter) last() uint64 {
+	return atomic.LoadUint64((*uint64)(c))
+}
+
+// residentManager creates and tracks cache residents.
+// It is also the source for resource identifiers in the cache.
+type residentManager struct {
+	residentCount *counter
+	resourceCount *counter
+
+	residents map[uint64]*resident
+}
+
+func newResidentManager() *residentManager {
+	residentC := counter(0)
+	resourceC := counter(0)
+
+	return &residentManager{
+		residentCount: &residentC,
+		resourceCount: &resourceC,
+		residents:     make(map[uint64]*resident),
+	}
+}
+
+// new creates a uniquely identified type-agnostic cache resident,
+// registers it in the internal map, then returns it.
+func (m *residentManager) new() *resident {
+	r := &resident{
+		id:      m.residentCount.next(),
+		manager: m,
+		workers: make(map[uint64]worker.Worker),
+	}
+	m.residents[r.id] = r
+	return r
+}
+
+func (m *residentManager) evict(id uint64) {
+	// TODO (manadart 2019-04-17): TBC when the mark/sweep logic is added.
+}
+
+func (m *residentManager) deregister(id uint64) {
+	delete(m.residents, id)
+}
+
+// resident is the base class for entities managed in the cache.
+type resident struct {
+	// id uniquely identifies this resident among all
+	// that were supplied by the same resident manager.
+	id uint64
+
+	// stale indicates that this cache resident is stale
+	// and is a candidate for removal.
+	stale bool
+
+	// manager is the resident manager that is responsible for this resource.
+	// If the resident is being evicted, it deregisters from the manager.
+	// TODO (manadart 2019-04-16): Maybe?
+	manager *residentManager
```

Comment:
Does the resident need the manager for anything other than deristering? If not, why not then just have a function here rather than a manager?

Reply:
Since it is also getting new worker IDs, perhaps another function to get those whose implementation just happens to be handled by the residentManager.